File: Algotithms/Others/rabin-karp.py
""" rabin karp implementation in python: """

d = 256 #number of characters in the input alphabet

def search(pattern, text, q):
    """ length of pattern"""
    m = len(pattern)
    """ length of text: """
    n = len(text)
    p = 0  # hash value for pattern
    t = 0   # hash value for text
    h = 1
    i = 0
    j = 0

    """The value of h would be "pow(d, M-1)%q" """
    for i in range(m - 1):
        h = (h * d) % q

    """ calculate the hash value of the pattern and first window of text: """
    for i in range(m):
        p = (d * p + ord(pattern[i])) % q
        t = (d * t + ord(text[i])) % q

    """ slide the pattern over text one by one:  """

    for i in range(n - m + 1):
        """
        Check the hash values of current window of text and
        pattern if the hash values match then only check
        for characters one by one
        """
        if p == t:
            """check for the characters one by one: """
            for j in range(m):
                if text[i + j] != pattern[j]:
                    break
            j += 1
            """if p == t and pat[0...M-1] = txt[i, i+1, ...i+M-1]"""
            if j == m:
                print("Pattern found at position: " + str(i + 1))
        """
        calculate the hash value of the next window of text. Remove the leading digit and 
        add trailing digits: 
        """
        if i < n - m:
            t = (d * (t - ord(text[i]) * h) + ord(text[i + m])) % q

            """we might get negative values of t, converting it into positive: """
            if t < 0:
                t = t + q

if __name__ == "__main__":
    text = "ABCCDDAEFG"
    pattern = "CDD"
    q = 13
    search(pattern, text, q)

File: Algorithms/Others/rabin-karp.cpp
/* rabin karp implentation in cpp: */
#include "bits/stdc++.h"

using namespace std;

#define d 256

void search(char * pat, char * txt, int q){
    int m = strlen(pat);
    int n = strlen(txt);
    int i, j;
    int p = 0; /* hash value for pattern*/
    int t = 0; /* hash value for txt*/
    int h = 1;

    /* the value of h would be "pow(d, M-1)%q"*/
    for (int i = 0; i < m - 1; i ++){
        h = (h * d) % q;
    }

    /* calculate the hash value of pattern and first window of text:*/
    for (i = 0; i < m; i ++){
        p = (d * p + pat[i]) % q;
        t = (d * t + txt[i]) % q;
    }

    /* slide the pattern over text one by one: */
    for (i = 0; i <= n - m; i++){
        /* check the hash values of current window of text and pattern. If they match
         then and only then do we check one by one: */
         if (p == t){
            /* check for the characters one by one: */
            for (j = 0; j < m; j ++){
                if (txt[i + j] != pat[j]){
                    break;
                }
            }
            /*if p == t and pat[0...M-1] = txt[i, i+1, ...i+M-1]*/

            if(j == m){
                cout << "Pattern found at index " << i << endl;
            }
         }

         /* calculate the hash value for the next window of text, remove the leading digit and add trailing digit*/
         if (i < n - m ){
            t = (d * (t - txt[i] * h) + txt[i + m]) % q;

            if (t < 0){
                t = t + q;
            }
         }
    }

}
/* driver code: */

int main(){
    char txt[] = "FOOD FIRST FOR FOOLS";
    char pat[] = "FOO";
    /* we mod to avoid overflowing value but we should take as big q as possible to avoid collision*/
    int q = INT_MAX;
    search(pat, txt,q);
    return 0;
}